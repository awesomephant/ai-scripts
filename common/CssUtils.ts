import roundTo from "./roundTo"

function formatCssRule(selector: string, obj: Record<string, string | number>): string {
	var css = selector + " {\r"
	for (var k in obj) {
		css += "\t" + k + ":" + obj[k] + ";\r"
	}
	css += "}\r"
	return css
}

function formatCssColor(r: number, g: number, b: number, a?: number) {
	if (a && a > 0 && a < 100) {
		return `rgba(${[r, g, b, roundTo(a * 0.01, 2)].join(",")})`
	}
	return `rgb(${[r, g, b].join(",")})`
}

function formatCssPct(part: number, whole: number, precision: number) {
	return roundTo((part / whole) * 100, precision) + "%"
}

export { formatCssColor, formatCssRule, formatCssPct }
