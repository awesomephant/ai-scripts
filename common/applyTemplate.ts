// Apply very basic string substitution to a template
export default function applyTemplate(template: string, replacements: any) {
	var keyExp = "([_a-zA-Z][\\w-]*)"
	var mustachePattern = new RegExp("\\{\\{\\{? *" + keyExp + " *\\}\\}\\}?", "g")
	var ejsPattern = new RegExp("<%=? *" + keyExp + " *%>", "g")
	var replace = function (match: string, name: string) {
		var lcname = name.toLowerCase()
		if (name in replacements) return replacements[name]
		if (lcname in replacements) return replacements[lcname]
		return match
	}
	// @ts-expect-error
	return template.replace(mustachePattern, replace).replace(ejsPattern, replace)
}
