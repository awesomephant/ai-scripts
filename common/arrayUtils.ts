function forEach(arr: any[], cb: (element: any, index: number) => any) {
	for (var i = 0, n = arr.length; i < n; i++) {
		cb(arr[i], i)
	}
}

function map(arr: any[], cb: (element: any, index: number) => any[]) {
	var arr2 = []
	for (var i = 0, n = arr.length; i < n; i++) {
		arr2.push(cb(arr[i], i))
	}
	return arr2
}

function filter(arr: any[], test: (element: any, index: number) => boolean) {
	var filtered = []
	for (var i = 0, n = arr.length; i < n; i++) {
		if (test(arr[i], i)) {
			filtered.push(arr[i])
		}
	}
	return filtered
}

// obj: value or test function
type ArrayTest = (element: any) => boolean | string | number

function indexOf(arr: any[], obj: ArrayTest): number {
	var test = typeof obj == "function" ? obj : null
	for (var i = 0, n = arr.length; i < n; i++) {
		if (test ? test(arr[i]) : arr[i] === obj) {
			return i
		}
	}
	return -1
}

function find(arr: any[], obj: ArrayTest): number | null {
	var i = indexOf(arr, obj)
	return i == -1 ? null : arr[i]
}

function contains(arr: any[], obj: ArrayTest): boolean {
	return indexOf(arr, obj) > -1
}

// alias for contains() with function arg
function some(arr: any[], cb: ArrayTest) {
	return indexOf(arr, cb) > -1
}

function forEachProperty(o: any[string], cb: (value: any, key: string) => any) {
	for (var k in o) {
		if (o.hasOwnProperty(k)) {
			cb(o[k], k)
		}
	}
}

function extend(o: any) {
	for (var i = 1; i < arguments.length; i++) {
		forEachProperty(arguments[i], add)
	}
	function add(v: any, k: string) {
		o[k] = v
	}
	return o
}

// Return new object containing properties of a that are missing or different in b
// Return null if output object would be empty
// a, b: JS objects
function objectDiff(a: any, b: any) {
	var diff: any = null
	for (var k in a) {
		if (a[k] != b[k] && a.hasOwnProperty(k)) {
			diff = diff || {}
			diff[k] = a[k]
		}
	}
	return diff
}

// return elements in array "a" but not in array "b"
function arraySubtract(a: any[], b: any[]) {
	var diff = [],
		alen = a.length,
		blen = b.length,
		i,
		j
	for (i = 0; i < alen; i++) {
		diff.push(a[i])
		for (j = 0; j < blen; j++) {
			if (a[i] === b[j]) {
				diff.pop()
				break
			}
		}
	}
	return diff
}

// Copy elements of an array-like object to an array
function toArray(obj: any[] | string) {
	var arr = []
	for (var i = 0, n = obj.length; i < n; i++) {
		// about 2x faster than push() (apparently)
		// arr.push(obj[i]);
		arr[i] = obj[i]
	}
	return arr
}

// multiple key sorting function based on https://github.com/Teun/thenBy.js
// first by length of name, then by population, then by ID
// data.sort(
//     firstBy(function (v1, v2) { return v1.name.length - v2.name.length; })
//     .thenBy(function (v1, v2) { return v1.population - v2.population; })
//     .thenBy(function (v1, v2) { return v1.id - v2.id; });
// );
type sortFn = (a: any, b: any) => number
function firstBy(f1: sortFn, f2: sortFn) {
	var compare = f2
		? function (a: any, b: any) {
				return f1(a, b) || f2(a, b)
		  }
		: f1
	// @ts-expect-error
	compare.thenBy = function (f) {
		return firstBy(compare, f)
	}
	return compare
}

export {
	forEach,
	find,
	map,
	filter,
	contains,
	some,
	extend,
	indexOf,
	forEachProperty,
	objectDiff,
	arraySubtract,
	toArray,
	firstBy
}
