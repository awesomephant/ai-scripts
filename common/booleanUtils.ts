/**
 * Coerce thruthy-sounding string to boolean
 * @param s
 * @returns boolean
 */
function isTrue(s: string | boolean): boolean {
	return s === "true" || s === "yes" || s === true
}

/**
 * Coerce falsy-sounding string to boolean true
 * @param s
 * @returns boolean
 */
function isFalse(s: string | boolean): boolean {
	return s === "false" || s === "no" || s === false
}

export { isTrue, isFalse }
