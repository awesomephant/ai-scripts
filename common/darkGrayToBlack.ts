export default function darkGrayToBlack(color: RGBColor, threshold: number = 36) {
	return color.red < threshold && color.green < threshold && color.blue < threshold
		? [0, 0, 0]
		: [color.red, color.green, color.blue]
}
