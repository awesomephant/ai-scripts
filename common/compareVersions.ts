import { map } from "./arrayUtils"

// assumes three-part version, e.g. 1.5.0
export default function compareVersions(va: string, vb: string) {
	const parse: (s: string, i: number) => number = (s, i) => {
		return parseInt(s)
	}
	const a: number[] = map(va.split("."), parse)
	const b: number[] = map(vb.split("."), parse)
	const diff = a[0] - b[0] || a[1] - b[1] || a[2] - b[2] || 0
	return (diff < 0 && -1) || (diff > 0 && 1) || 0
}
