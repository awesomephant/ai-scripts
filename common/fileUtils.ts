import { ai2HTMLSettings } from "../ai2html/types"
import { forEach } from "./arrayUtils"
import { parseYaml } from "./yamlUtils"

function folderExists(path: string) {
	return new Folder(path).exists
}

function fileExists(path: string) {
	return new File(path).exists
}

function deleteFile(path: string) {
	var file = new File(path)
	if (file.exists) {
		file.remove()
	}
}
function readFile(fpath: string, onerror?: (err: string) => any, enc?: string) {
	var content = null
	var file = new File(fpath)
	if (file.exists) {
		if (enc) {
			file.encoding = enc
		}
		file.open("r")
		if (file.error && onerror) {
			// (on macOS) restricted permissions will cause an error here
			onerror("Unable to open " + file.fsName + ": [" + file.error + "]")
			return null
		}
		content = file.read()
		file.close()
		// (on macOS) 'file.length' triggers a file operation that returns -1 if unable to access file
		if (!content && (file.length > 0 || file.length == -1) && onerror) {
			onerror("Unable to read from " + file.fsName + " (reported size: " + file.length + " bytes)")
		}
	} else {
		if (onerror) {
			onerror(fpath + " could not be found.")
		}
	}
	return content
}

function readTextFile(fpath: string): string {
	// This function used to use File#eof and File#readln(), but
	// that failed to read the last line when missing a final newline.
	return readFile(fpath, () => {}, "UTF-8") || ""
}

function saveTextFile(dest: string, contents: string) {
	var fd = new File(dest)
	fd.open("w", "TEXT", "TEXT")
	fd.lineFeed = "Unix"
	fd.encoding = "UTF-8"
	fd.writeln(contents)
	fd.close()
}

function checkForOutputFolder(
	folderPath: string,
	nickname: string,
	oncreated: (message: string) => void,
	onerror: (err: string) => void
) {
	const outputFolder = new Folder(folderPath)
	if (!outputFolder.exists) {
		const outputFolderCreated = outputFolder.create()
		if (outputFolderCreated) {
			if (oncreated) {
				oncreated("The " + nickname + " folder did not exist, so the folder was created.")
			}
		} else {
			if (onerror) {
				onerror("The " + nickname + " folder did not exist and could not be created.")
			}
		}
	}
}

function readJsonFile(fpath: string, JSON: any, onerror: (e: string) => any) {
	var content = readTextFile(fpath)
	var json = null
	if (!content) {
		// removing for now to avoid double warnings
		// warn('Unable to read contents of file: ' + fpath);
		return {}
	}
	try {
		json = JSON.parse(content)
	} catch (e: any) {
		if (onerror) {
			onerror("Error parsing JSON from " + fpath + ": [" + e.message + "]")
		}
	}
	return json
}

function readYamlConfigFile(path: string, JSON: any): any | null {
	return fileExists(path) ? parseYaml(readTextFile(path), JSON) : null
}

// Similar to Node.js path.join()
function pathJoin(...args: string[]) {
	var path = ""
	forEach(args, function (arg, i) {
		if (!arg) return
		arg = String(arg)
		// Drop leading slash, except on the first argument  because that's
		// necessary to differentiate different volumes on Windows
		if (i > 0) {
			arg = arg.replace(/^\/+/, "")
		}
		arg = arg.replace(/\/+$/, "")

		if (path.length > 0) {
			path += "/"
		}
		path += arg
	})
	return path
}

// Split a full path into directory and filename parts
function pathSplit(path: string) {
	var parts = path.split("/")
	var filename = parts.pop()
	return [parts.join("/"), filename]
}

function getScriptDirectory() {
	return new File($.fileName).parent
}
function getImageFolder(settings: ai2HTMLSettings, docPath: string) {
	// return pathJoin(docPath, settings.html_output_path, settings.image_output_path);
	return pathJoin(docPath, settings.image_output_path)
}

export {
	folderExists,
	fileExists,
	deleteFile,
	pathJoin,
	pathSplit,
	readTextFile,
	saveTextFile,
	getScriptDirectory,
	readFile,
	readJsonFile,
	readYamlConfigFile,
	checkForOutputFolder,
	getImageFolder
}
