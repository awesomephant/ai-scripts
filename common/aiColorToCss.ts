import { formatCssColor } from "./cssUtils"
import darkGrayToBlack from "./darkGrayToBlack"

import type { Color, RGBColor, SpotColor, GrayColor, NoColor, CMYKColor } from "../ai2html/types"

interface aiColorToCssResult {
	color: string
	warning: string | null
}

/**
 * Transforms an Illustrator object into a CSS color string
 * @param opacity (optional): opacity [0-100]
 * @returns
 */

export default function aiColorToCss(
	color: RGBColor | SpotColor | GrayColor | NoColor | CMYKColor,
	opacity: number = 100
): aiColorToCssResult {
	if (color.typename === "SpotColor") {
		return aiColorToCss(color.spot.color, opacity)
	}
	if (color.typename === "RGBColor") {
		const [r, g, b] = darkGrayToBlack(color)
		return {
			color: formatCssColor(r, g, b, opacity),
			warning: null
		}
	} else if (color.typename === "GrayColor") {
		const v = Math.round(((100 - color.gray) / 100) * 255)
		return {
			color: formatCssColor(v, v, v, opacity),
			warning: null
		}
	} else if (color.typename === "NoColor") {
		// warnings are processed later, after ranges of same-style chars are identified
		// TODO: add text-fill-specific warnings elsewhere
		return {
			color: formatCssColor(0, 255, 0, 100),
			warning:
				'The text "%s" has no fill. Please fill it with an RGB color. It has been filled with green.'
		}
	}
	return {
		color: formatCssColor(0, 0, 0, opacity),
		warning: `The text "%s" has ${color.typename} fill. Please fill it with an RGB color.`
	}
}
