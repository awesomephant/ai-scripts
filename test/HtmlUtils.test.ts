import { describe, it, expect, vi } from "vitest"
import { cleanHtmlTags, findHtmlTag, injectCSSinSVG } from "../common/htmlUtils"

describe("cleanHtmlTags()", () => {
	it("converts smart quotes to double quotes", function () {
		expect(cleanHtmlTags("<a href=”#”>link text</a>")).toBe('<a href="#">link text</a>')
	})
	it("warns on formatting tags", () => {
		const ontagfound = vi.fn()
		cleanHtmlTags('should <b>error</b>"', ontagfound)
		expect(ontagfound).toHaveBeenCalled()
		expect(ontagfound).toHaveBeenCalledWith(
			"Found a <b> tag. Try using Illustrator formatting instead."
		)
	})
})

describe("findHtmlTag()", function () {
	it("finds name of HTML tag in a string", function () {
		expect(findHtmlTag('<a href="#">link text')).toBe("a")
	})
})

describe("injectCSSinSVG()", function () {
	it("injects a <style> element at the end of an <svg> element", function () {
		expect(
			injectCSSinSVG("<svg><rect x1='1' x2='5'></rect></svg>", "rect {fill: red}")
		).toMatchSnapshot()
	})
})
