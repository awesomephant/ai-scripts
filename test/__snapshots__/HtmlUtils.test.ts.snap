// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`injectCSSinSVG() > injects a <style> element at the end of an <svg> element 1`] = `"<svg><rect x1='1' x2='5'></rect><style>rect {fill: red}</style></svg>"`;
